### ISSUE TITLE
Backslash in file name

### ISSUE THREAD

**Issue by mikkleini** [gcovr/gcovr#67]
Please update lines 1472 and 1477 in scripts/gcovr which assemble the cdata._sourcefile string (function print_html_report). Today code replaces Unix slash with underscore but not the Windows backslash and colon characters so gcovr doesn't work. I have updated my local copy and with this small change it works very well with Windows (unline LCov which requires much more updates :( ).

Edit: maybe it's even nicer to add a separate function which creates OS safe file name?

**Comment by mrx23dot** [comment]
Please use the following for all inputs:

from os.path import normpath
abs_filename = os.path.abspath(normpath(filename))

This will unify all paths for given OS.

**Comment by mrx23dot** [comment]
I use:

-# creates folders, normalizes path
def CheckPath(path_file):
    tmp_path = os.path.abspath(os.path.normpath(path_file))

    if os.path.isfile(tmp_path):
      path, name = os.path.split(tmp_path); # os.path.dirname no reliable
    else:
      path = tmp_path;

    -# create all dirs
    if not os.path.exists(path) :
      os.makedirs(path); # recursive
    return tmp_path

**Comment by Sildra** [comment]
I don't think that is the right thing to handle paths because underscores can be part of the filename and when the transformation is done there is no way to differentiate a folder from what was part of the filename.
Instead Gcovr should recreate the architecture of the project :
```
<Folder>
|-- <File1>.cc
|-- <File2>.cc
```
invocation:
```
$ g++ --coverage...
$ ./a.out
$ gcovr -r . --xml --html --html-details -o _cov
```
output:
```
_cov
|--index.html
|--coverage.xml
|--<Folder>
  |-- <File1>.cc.html
  |-- <File2>.cc.html
```
diff (with comments):
```
--- gcovr.old	2015-06-16 15:55:16.964311300 +0200
+++ gcovr	2015-06-16 16:19:09.908591400 +0200
@@ -1470,2 +1470,4 @@
### Use relative path by default (absolute path messes with the browser under Windows)
-        ttmp = os.path.abspath(options.output).split('.')
-        if len(ttmp) > 1:
+        ttmp = options.output.split('.')
+        if options.output.split('.')[-1] != "html":
### Fuck windows and his separator, enforce Unix style because it works under Windows
### the other way does not work
+            cdata._sourcefile = cdata._filename.replace('\\','/') + '.html'
+        elif len(ttmp) > 1:
@@ -1539 +1541,5 @@
### If root is not html file we assume this is a directory and we create it
-        OUTPUT = open(options.output, 'w')
+        if options.output.split('.')[-1] != "html":
+            os.makedirs(options.output, exist_ok=True)
+            OUTPUT = open(os.path.join(options.output, "index.html"), 'w')
+        else:
+        	OUTPUT = open(options.output, 'w')
### ensure folder exists
@@ -1599 +1605,2 @@
-        OUTPUT = open(cdata._sourcefile, 'w')
+        os.makedirs(os.path.dirname(os.path.join(options.output, cdata._sourcefile)), exist_ok=True)
+        OUTPUT = open(os.path.join(options.output, cdata._sourcefile), 'w')
### changes in --html-absolute-path
@@ -1632,2 +1639,2 @@
-    if options.relative_anchors:
-        sourcefile = os.path.basename(sourcefile)
+    if not options.relative_anchors:
+        sourcefile = os.path.abspath(sourcefile)
### If root is not xml file we assume this is a directory and we create it
@@ -1898 +1905,5 @@
-        OUTPUT = open(options.output, 'w')
+        if options.output.split('.')[-1] != "xml":
+            os.makedirs(options.output, exist_ok=True)
+            OUTPUT = open(os.path.join(options.output, "coverage.xml"), 'w')
+        else:
+        	OUTPUT = open(options.output, 'w')
### Allow --xml and --html
@@ -2179 +2190 @@
-elif options.html:
+if options.html:
@@ -2181 +2192 @@
-else:
+if not (options.xml or options.prettyxml or options.html):
```
When starting the SimpleHTTPServer from the _cov directory or openning the index.html, firefox can open others files.

**Comment by Spacetown** [comment]
Solved with #375.
